clients/pkg/promtail/targets/objectstore/reader.go

```diff
@@ -0,0 +1,251 @@
+package objectstore
+
+import (
+	"bufio"
+	"bytes"
+	"compress/gzip"
+	"context"
+	"fmt"
+	"io"
+	"strconv"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/go-kit/log"
+	"github.com/go-kit/log/level"
+	"github.com/pkg/errors"
+	"github.com/prometheus/common/model"
+	"go.uber.org/atomic"
+
+	"github.com/grafana/loki/clients/pkg/promtail/api"
+	"github.com/grafana/loki/clients/pkg/promtail/positions"
+
+	"github.com/grafana/loki/pkg/logproto"
+	"github.com/grafana/loki/pkg/storage/chunk"
+)
+
+type objectReader struct {
+	metrics      *Metrics
+	storeName    string
+	handler      api.EntryHandler
+	objectClient chunk.ObjectClient
+	ackMessage   ackMessage
+	logger       log.Logger
+	readerMtx    sync.RWMutex
+	labels       model.LabelSet
+	positions    positions.Positions
+	object       *promtailObject
+	active       *atomic.Bool
+	ctx          context.Context
+	cancel       context.CancelFunc
+	wg           sync.WaitGroup
+}
+
+type promtailObject struct {
+	Key        string
+	ModifiedAt time.Time
+	BytesRead  int64
+}
+
+func newObjectReader(
+	metrics *Metrics,
+	storeName string,
+	object chunk.StorageObject,
+	ackMessage ackMessage,
+	objectClient chunk.ObjectClient,
+	logger log.Logger,
+	labels model.LabelSet,
+	positions positions.Positions,
+	handler api.EntryHandler,
+	pos int64) *objectReader {
+	ctx, cancel := context.WithCancel(context.Background())
+	objectReader := &objectReader{
+		metrics:      metrics,
+		storeName:    storeName,
+		logger:       logger,
+		labels:       labels,
+		positions:    positions,
+		handler:      handler,
+		objectClient: objectClient,
+		ackMessage:   ackMessage,
+		active:       atomic.NewBool(true),
+		ctx:          ctx,
+		cancel:       cancel,
+		object: &promtailObject{
+			Key:        object.Key,
+			ModifiedAt: object.ModifiedAt,
+			BytesRead:  pos,
+		},
+	}
+
+	go objectReader.run()
+	metrics.objectActive.Add(1.)
+	return objectReader
+}
+
+func (r *objectReader) run() {
+	level.Info(r.logger).Log("msg", "start reading object", "object", r.object.Key)
+	r.wg.Add(1)
+	positionSyncPeriod := r.positions.SyncPeriod()
+	positionWait := time.NewTicker(positionSyncPeriod)
+	defer func() {
+		positionWait.Stop()
+		r.wg.Done()
+	}()
+
+	bufReader, err := r.GetObject()
+	if err != nil {
+		level.Error(r.logger).Log("msg", "error in fetching and initializing object reader", "err", err)
+		r.active.Store(false)
+		return
+	}
+
+	// add s3_object label
+	labels := r.labels.Merge(model.LabelSet{
+		model.LabelName("s3_object"): model.LabelValue(r.object.Key),
+	})
+
+	for {
+		select {
+		case <-r.ctx.Done():
+			return
+		case <-positionWait.C:
+			r.markPositionAndSize()
```

Comment:
I don't think you need to have a ticker to mark the position, you could do it on every line.

Reply:
Okay